asr1k_neutron_l3/plugins/l3/schedulers/simple_asr1k_scheduler.py

```diff
@@ -43,12 +45,27 @@ def _get_candidates(self, plugin, context, sync_router):
                            'agent_id': current_l3_agents[0]['id']})
                 return []
 
-            candidates = plugin.get_l3_agents(context, active=True)
+            orig_candidates = plugin.get_l3_agents(context, active=True)
+
+            # router creation with az hint: only schedule on agent with appropriate AZ
+            # router creation without az hint: only schedule on agent with no AZ
+            az_hints = orig_az_hints = self._get_az_hints(sync_router)
+            if not az_hints or az_hints[0] in asr1k_const.NO_AZ_LIST:
+                az_hints = asr1k_const.NO_AZ_LIST
+            candidates = [c for c in orig_candidates if c.availability_zone in az_hints]
+
+            if not candidates and orig_az_hints and cfg.CONF.asr1k.ignore_invalid_az_hint_for_router:
+                LOG.warning("No candidate found for router %s with az hint %s, reverting to original candidate list %s",
+                            sync_router['id'], az_hints, orig_candidates)
+                candidates = orig_candidates
+
             if not candidates:
-                LOG.warning('No active L3 agents')
+                LOG.warning('No active L3 agents found for router %s (az hints were %s)',
+                            sync_router['id'], orig_az_hints)
```

Comment:
I assume the router precommit hook executed before scheduling? Otherwise we silently omit scheduling which I don't think is necessarily what we want.